### PR TITLE
feat: users can save their progress

### LINKED ISSUE #17609: Users are able to manually save the current state of their form
### Description

Users are given the ability to save the current state of their form without submitting it.

### Acceptance Criteria
- [ ] Tasks MUST show a control which allows users to manually save the contents of their form as a draft
- [ ] Tasks MUST restore the form to its previously saved state when loaded
- [ ] Tasks MUST allow the user to save their progress at any time **ignoring any validation**
- [ ] Tasks MUST show a visual indication when the form is in the progress of being saved
- [ ] Tasks MUST show a notification when the form is saved successfully
- [ ] Tasks MUST show a warning if saving fails
- [ ] Tasks MUST NOT allow a user to save a task when they are not assigned to the task
- [ ] Tasklist MUST track when the Save button has been clicked

#### UX
https://www.figma.com/file/sL36etbFEfaZ3Y1a2NDOY0/Tasklist-Task-Details?type=design&node-id=147-16247&mode=design&t=KRUQi0oOnpoZxS7E-0

image
image
image
image

### PR DESCRIPTION
**Pull request by douglasbouttell-camunda**
## Description

Some tasks require significant effort from the user or due to circumstances tasks may need to be transferred to other people. Camunda should support allowing users to save their progress in order to help their job.

## Related issues

closes #17609

### CODE REVIEW

File: tasklist/client/src/Tasks/Task/Variables/index.tsx [hunk 1]
```diff
@@ -54,6 +60,56 @@ type Props = {
   user: CurrentUser;
 };
 
+function createInitialFormValuesFromVariables(variables: Variable[]) {
+  const entries: Array<[string, string | null]> = [];
+  const newVariables: Array<Pick<Variable, 'name' | 'value'>> = [];
+
+  variables.forEach((variable) => {
+    if (variable.draft) {
```
Comment (vsgoulart) [inline tasklist/client/src/Tasks/Task/Variables/index.tsx:68]: Let's invert the logic to use a explicit check like everywhere else

```suggestion
    if (variable.draft === null) {
```

File: tasklist/client/src/Tasks/Task/FormJS/index.tsx [hunk 2]
```diff
@@ -169,7 +190,30 @@ const FormJS: React.FC<Props> = ({
...
+        >
+          <SaveButton
+            savingState={savingState}
+            onClick={() => {
+              save(
+                formatFormDataToVariables(
+                  formManagerRef.current
+                    ?.get('form')
+                    ._getSubmitData() as object,
+                ),
```
Comment (vsgoulart) [inline tasklist/client/src/Tasks/Task/FormJS/index.tsx:211]: We shouldn't be using `form-js` internal methods like this, part of the epic should have been changing adapting `form-js` and make this public...

**Review by vsgoulart — COMMENTED** [review]
@douglasbouttell-camunda This PR is already a bit too, next time please it into smaller ones (a PR for the visual component and one for the mutation logic for example)

File: tasklist/client/src/Tasks/Task/Variables/index.tsx [hunk 3]
```diff
@@ -244,7 +275,16 @@ const Variables: React.FC<Props> = ({
...
-                <DetailsFooter>
+                <DetailsFooter
+                  className={styles.actionBar}
+                  status={
+                    savingState === 'finished' ? (
+                      <SaveSuccessMessage />
+                    ) : savingState === 'error' ? (
+                      <SaveFailedMessage />
+                    ) : undefined
+                  }
```
[the remark this answers is not in the document]
Comment (douglasbouttell-camunda) [reply tasklist/client/src/Tasks/Task/Variables/index.tsx:286]: the AsyncActionButton doesn't do what the UX design wants.

**Review by vsgoulart — APPROVED** [review]

**Comment by vsgoulart** [comment]
@esraagamal6 Could you please test this PR?

**Review by vsgoulart — APPROVED** [review]

**Comment by esraagamal6** [comment]
due to architecture decisions on the new API, we will not publish this feature with this implementation.
cc: @vsgoulart @marcosgvieira

**Pull request closed without merging**